core/ir/category_api.h

```diff
@@ -0,0 +1,15 @@
+#ifndef _DR_IR_CATEGORIES_H_
+#define _DR_IR_CATEGORIES_H_ 1
+
+typedef enum _dr_instr_category_type_t {
+    DR_INSTR_CATEGORY_UNCATEGORIZED = 0x0, /* Uncategorised. */
+    DR_INSTR_CATEGORY_MATH_INT = 0x1,      /* Integer arithmetic operations */
+    DR_INSTR_CATEGORY_MATH_FLOAT = 0x2,    /* Floating-Point arithmetic operations */
+    DR_INSTR_CATEGORY_LOAD = 0x4,          /* Loads */
+    DR_INSTR_CATEGORY_STORE = 0x8,         /* Stores */
+    DR_INSTR_CATEGORY_BRANCH = 0x10,       /* Branches */
+    DR_INSTR_CATEGORY_SIMD = 0x20,         /* Operations with vector registers (SIMD) */
+    DR_INSTR_CATEGORY_OTHER = 0x40         /* Other types of instructions */
```

Comment:
Xref #6238 where we also had a list of categories, for the Google workload traces where we can't give out opcodes (but we'd want a programmatic way to find those categories from a real opcode, so this PR would fit the bill, if on all arches).

Comparing the category lists: looks like yours is also composable which is good.  We wanted to distinguish single and double precision floating point, and to separate out FMA (fused multiply add).  Maybe those are the main differences between the two lists; the extra categories in #6238 are less important.

Reply:
> Xref #6238 where we also had a list of categories, for the Google workload traces where we can't give out opcodes (but we'd want a programmatic way to find those categories from a real opcode, so this PR would fit the bill, if on all arches).
> 
> Comparing the category lists: looks like yours is also composable which is good. We wanted to distinguish single and double precision floating point, and to separate out FMA (fused multiply add). Maybe those are the main differences between the two lists; the extra categories in #6238 are less important.

Could not find something like this for x86_64. We create just flat list for opcodes about their categories.
We could start from this list and add new ones categories. For example, some instructions could be load ans SIMD.
So, float could be in DR_INSTR_CATEGORY_MATH_FLOAT |  DR_INSTR_CATEGORY_MATH_SINGLE_FLOAT or  DR_INSTR_CATEGORY_MATH_FLOAT |  DR_INSTR_CATEGORY_MATH_DOUBLE_FLOAT

Reply:
There should be combinations today: a scatter instruction should be STORE|SIMD and gather LOAD|SIMD.  If that's not simple through the encoding bit scheme, maybe these would have to be flags added to every opcode in the codec*.txt files.  That is likely how it would be done for x86 and AArch32 as well: flags on every encoding entry (or else it has a be a big external switch statement on opcodes like today's instr_is_floating_ex()).

Reply:
That was done already over bits
        category =
            (BITS(enc, 22, 22) == 0) ? DR_INSTR_CATEGORY_STORE : DR_INSTR_CATEGORY_LOAD;
        if ((op0 & 0xc) == 0 && BITS(enc, 26, 26) == 1)
            category |= DR_INSTR_CATEGORY_SIMD;

Reply:
OK.  Sure, let's start with this, and if we want to add single vs double, FMA, atomics, memory barriers, etc. later they can be appended.